In [43]:
# Import data
import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\Anwender\Desktop\29-11-2020 Instacart Basket Analysis'

In [45]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [44]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [58]:
# Importing orders_wrangled.csv from prepared data folder 
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [12]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# CHECKING FOR MIXED - TYPE DATA

In [13]:
# Create a dataframe
df_test = pd.DataFrame()

In [15]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [16]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [17]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [87]:
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col +' has mixed-type data')
  else:
    print (col + ' data type is uniform')
  

product_id data type is uniform
product_name has mixed-type data
aisle_id data type is uniform
department_id data type is uniform
prices data type is uniform


In [88]:
# Change to string
df_prods['product_name'] = df_prods['product_name'].astype('str')

In [89]:
df_prods['product_name'].astype('string')

0                               Chocolate Sandwich Cookies
1                                         All-Seasons Salt
2                     Robust Golden Unsweetened Oolong Tea
3        Smart Ones Classic Favorites Mini Rigatoni Wit...
4                                Green Chile Anytime Sauce
                               ...                        
49688            Vodka, Triple Distilled, Twist of Vanilla
49689                   En Croute Roast Hazelnut Cranberry
49690                                     Artisan Baguette
49691           Smartblend Healthy Metabolism Dry Cat Food
49692                               Fresh Foaming Cleanser
Name: product_name, Length: 49693, dtype: string

In [90]:
# Check the data type to see if it executed 
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49693 entries, 0 to 49692
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49693 non-null  int64  
 1   product_name   49693 non-null  object 
 2   aisle_id       49693 non-null  int64  
 3   department_id  49693 non-null  int64  
 4   prices         49693 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [18]:
# Convert column name, you can also change it to numeric by imputing this fucntion in the str 0- int64
# Fix the mixed-type column
df_test['mix'] = df_test['mix'].astype('str')

In [19]:
# 3, Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [59]:
# Creating a subset with the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [91]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
df_prods['product_name'].value_counts(dropna = False)

NaN                                                     16
Black House Coffee Roasty Stout Beer                     2
Adore Forever Body Wash                                  2
Gluten Free Organic Peanut Butter & Chocolate Cereal     2
Fiber 4g Gummy Dietary Supplement                        2
                                                        ..
Kosher Dill Mini Pickles                                 1
Chili Rice Scooter                                       1
All Natural Chicken Apple Sausage                        1
Quartered Artichoke Hearts                               1
Fresh Foaming Cleanser                                   1
Name: product_name, Length: 49673, dtype: int64

In [61]:
# Creating a new data frame with clean data
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [62]:
# Compare the number of rows in your current dataframe with the number in your subset once the missing rows have been removed:
df_prods.shape

(49693, 5)

In [63]:
df_prods_clean.shape

(49677, 5)

In [40]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [64]:
df_prods_clean.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

# Looking for duplicates

In [47]:
df_ords.shape

(3421083, 7)

In [49]:
df_prods.shape

(49693, 5)

In [50]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [51]:
# Checking the number of rows in the prods_clean data frame
df_prods_clean.shape

(49677, 5)

In [54]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [55]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [56]:
# Drop duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [57]:
# Checking the number of rows in the data frame without duplicates 
df_prods_clean_no_dups.shape

(49672, 5)

# QUESTIONS AND SOLUTION 

In [70]:
# 2, Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
# Tip: Keep an eye on min and max values!
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

looking at the max value from the price column which is 99999, we can see is a high numeric value. this value needs to be checked to know if it's actually correct or a error.

In [92]:
# 3, Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col + ' has mixed type data')
  else:
    print (col + ' data type is uniform')

Unnamed: 0 data type is uniform
order_id data type is uniform
user_id data type is uniform
order_number data type is uniform
orders_day_of_week data type is uniform
order_hour_of_day data type is uniform
days_since_prior_order data type is uniform
unknown_order data type is uniform


In [ ]:
# 4, If you find mixed-type data, fix it. The column in question should contain observations of a single data type.
No mixed-type columns in the df_ords data frame

In [73]:
# 5, Run a check for missing values in your df_ords dataframe.
# In a markdown cell, report your findings and propose an explanation for any missing values you find.
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [74]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [75]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

looking at the output we can see the days_since_prior_order has a lot of missing values this could be as a result of customers orders not accounted or been recorded due to previous orders without subscriptions which could be probably their first order.

In [ ]:
# 6, Address the missing values using an appropriate method.
# In a markdown cell, explain why you used your method of choice.

information are gotten from the data so there isn't need of removing the data, by so doing we need to Create a new variable that acts like a flag based on the missing value. new column can also be created to indicate these values to see what it entails which explain the similarity of the days_since_prior_order.

In [79]:
df_ords_clean = df_ords

In [93]:
df_ords_clean ['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [95]:
df_ords_clean ['first_order']

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Name: first_order, Length: 3421083, dtype: bool

In [96]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  unknown_order  first_order  
0                  8                     NaN           True         True  
1                  7                    15.0          False        False  
2                 12                    21.0          False        False  
3                  7                    29.0          False        False  
4                 15                    28.0          False        False

In [97]:
df_ords_clean.tail()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  unknown_order  first_order  
3421078                 18                    29.0          False        False  
3421079                 10                    30.0          False        False  
3421080                 12                    18.0          False        False  
3421081                 12                     7.0          False        False  
3421082                 14                    30.0          False        False

In [ ]:
# another method to flag
df_ords['first_order'] = np.where(np.isnan(df_ords['days_since_prior_order'].values), True, False)

In [83]:
# 7, Run a check for duplicate values in your df_ords data.
# In a markdown cell, report your findings and propose an explanation for any duplicate values you find.
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [84]:
# check duplicate
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, unknown_order]
Index: []

after checking for duplicate i notice the output was empty

In [85]:
# 8, Address the duplicates using an appropriate method.
# In a markdown cell, explain why you used your method of choice.
df_ords_dups.shape

(0, 8)

In [100]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_cleanednew.csv'))

In [101]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checkednew.csv'))